File: 2021/07/7.py
"""https://adventofcode.com/2021/day/7."""


def intrange(n):
    """Get sum of integer range."""
    return int(n*(n+1)/2)


def part1(data):
    """Solution for first puzzle.

    Sum distance of all positions to median.
    """
    positions = sorted(list(map(int, data.split(','))))
    median = positions[(len(positions)) // 2]
    return sum([abs(median - position) for position in positions])


def part2(data):
    """Solution to second puzzle.

    Brute force solution with checking all combinations.
    """
    fuel_spent = list()
    positions = list(map(int, data.split(',')))
    for target in range(min(positions), max(positions)+1):
        fuel_spent.append(sum([intrange(abs(target-position)) for position in positions]))
    return min(fuel_spent)


if __name__ == "__main__":
    with open('data/7.in') as f:
        data = f.read()

    print(f'Part 1: {part1(data)}')
    print(f'Part 2: {part2(data)}')

File: 2021/07/7.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#include "adoc.h"

#define INPUT_FILENAME "data/7.in"


void part1(std::vector<int> positions)
{
    std::sort(positions.begin(), positions.end());
    int median = positions.at(positions.size() / 2);

    int fuel = 0;
    for (auto p = positions.begin(); p != positions.end(); p++)
    {
        fuel += abs(median - *p);
    }
    std::cout << "Part 1: " << fuel << std::endl;
}

void part2(std::vector<int> positions)
{
    std::vector<int> fuel_spent;
    std::sort(positions.begin(), positions.end());

    for (auto target = positions.begin(); target != positions.end(); target++)
    {
        int fuel = 0;
        for (auto p = positions.begin(); p != positions.end(); p++) 
        {
            int diff = abs(*target - *p);
            fuel += (diff*(diff+1))/2;
        }
        fuel_spent.push_back(fuel);
    }

    auto result = *std::min_element(fuel_spent.begin(), fuel_spent.end());
    std::cout << "Part 2: " << result << std::endl;
}


int main() 
{
    std::vector<std::string> lines = adoc::read_lines_from_file(INPUT_FILENAME);
    std::vector<std::string> integer_strings = adoc::split_string(&lines.at(0), ",");
    std::vector<int> positions = adoc::convert_string_list_to_integer(&integer_strings);

    part1(positions);
    part2(positions);
}
